src/charm.py

```
@@ -380,6 +381,36 @@ def _get_nginx_pebble_config(self, _) -> ops.pebble.LayerDict:
         }
         return typing.cast(ops.pebble.LayerDict, layer)
 
+    def _get_redis_cache_url(self) -> str:
+        """Get Url for redis-cache charm.
+
+        Returns:
+            Url for the redis-cache charm.
+        """
+        relation = self.model.get_relation(self.redis_cache.relation_name)
+        if relation:
+            relation_data = relation.data[self.model.relations["redis-cache"][0].app]
+            ip = relation_data["leader-host"]
+            unit_relation = self.redis_cache.relation_data
+            port = unit_relation["port"]
+            return f"redis://{ip}:{port}"
+        return ""
+
+    def _get_redis_broker_url(self) -> str:
+        """Get Url for redis-broker charm.
+
+        Returns:
+            Url for the redis-broker charm.
+        """
+        relation = self.model.get_relation(self.redis_broker.relation_name)
+        if relation:
+            relation_data = relation.data[self.model.relations["redis-broker"][0].app]
+            ip = relation_data["leader-host"]
+            unit_relation = self.redis_broker.relation_data
+            port = unit_relation["port"]
+            return f"redis://{ip}:{port}"
+        return ""
```

It seems like it might be worth refactoring this to a single function that you pass the relation name to?

---

There"s definitely code duplication here

---

Thank you! Turned it into 1 function with redis name as input.

Should I return "" when redis name is wrong or throw an exception?

```python
    def _get_redis_url(self, redis_name: str) -> str:
        """Get Url for redis charm.

        Args:
            redis_name (str): Name of the redis charm to connect to.

        Returns:
            Url for the redis charm.
        """
        if redis_name == "redis-broker":
            redis = self.redis_broker
        elif redis_name == "redis-cache":
            redis = self.redis_cache
        else:
            return ""

        relation = self.model.get_relation(redis.relation_name)
        if relation:
            relation_data = relation.data[self.model.relations[redis_name][0].app]
            ip = relation_data["leader-host"]
            unit_relation = redis.relation_data
            port = unit_relation["port"]
            return f"redis://{ip}:{port}"
        return ""
```

---

I think you should raise an exception, because it would definitely be unexpected